File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples/VB/WPFTemplateApp/WPFTemplateApp/DataClasses/ProductManager.vb
Imports System
Imports System.Collections.ObjectModel
Imports System.Diagnostics
Imports System.IO
Imports System.Linq
Imports System.Xml.Linq

Public Class ProductManager
#Region "GetProducts Methods"
  Public Function GetProducts() As ObservableCollection(Of Product)
    Dim ret As ObservableCollection(Of Product)
    Dim fileName As String

    Try
      fileName = WPFCommon.GetCurrentDirectory() & "\Xml\Product.xml"

      If File.Exists(fileName) Then
        ret = GetProducts(fileName)
      Else
        ret = GetProductMocks()
      End If
    Catch ex As Exception
      Debug.WriteLine(ex.ToString())
      ' Still return some data
      ret = GetProductMocks()
    End Try

    Return ret
  End Function

  Protected Function GetProducts(ByVal fileName As String) As ObservableCollection(Of Product)
    Dim ret As ObservableCollection(Of Product)
    Dim elem As XElement = XElement.Load(fileName)

    Dim items = From prod In elem.Descendants("Product")
                Select New Product() With { _
     .ProductId = Convert.ToInt32(prod.Element("ProductId").Value), _
     .ProductName = prod.Element("ProductName").Value, _
     .Price = Convert.ToDecimal(prod.Element("Price").Value), _
     .IsActive = Convert.ToBoolean(prod.Element("IsActive").Value) _
    }

    ret = New ObservableCollection(Of Product)(items)

    Return ret
  End Function
#End Region

#Region "Insert Method"
  Public Function Insert(ByVal entity As Product) As Integer
    Dim ret As Integer = 0

    ' TODO: Insert Data Here

    Return ret
  End Function
#End Region

#Region "Update Method"
  Public Function Update(ByVal entity As Product) As Integer
    Dim ret As Integer = 0

    ' TODO: Update Data Here

    Return ret
  End Function
#End Region

#Region "Delete Method"
  Public Function Delete(ByVal entity As Product) As Integer
    Dim ret As Integer = 0

    ' TODO: Delete Data Here

    Return ret
  End Function
#End Region

#Region "Mock Data"
  Protected Overridable Function GetProductMocks() As ObservableCollection(Of Product)
    Dim entity As Product
    Dim ret As New ObservableCollection(Of Product)()

    entity = New Product()
    entity.ProductId = 1
    entity.ProductName = "(Mock) Product 1"
    entity.Price = 100
    entity.IsActive = True
    ret.Add(entity)

    entity = New Product()
    entity.ProductId = 2
    entity.ProductName = "(Mock) Product 2"
    entity.Price = 200
    entity.IsActive = True
    ret.Add(entity)

    entity = New Product()
    entity.ProductId = 3
    entity.ProductName = "(Mock) Product 3"
    entity.Price = 300
    entity.IsActive = False
    ret.Add(entity)

    Return ret
  End Function
#End Region
End Class

File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples/CSharp/WPFTemplateApp/WPFTemplateApp/DataClasses/ProductManager.cs
using System;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Xml.Linq;

namespace WPFTemplateApp
{
  public class ProductManager
  {
    #region GetProducts Methods
    public ObservableCollection<Product> GetProducts()
    {
      ObservableCollection<Product> ret;
      string fileName;

      try
      {
        fileName = WPFCommon.GetCurrentDirectory() +
          @"\Xml\Product.xml";

        if (File.Exists(fileName))
          ret = GetProducts(fileName);
        else
          ret = GetProductMocks();
      }
      catch (Exception ex)
      {
        Debug.WriteLine(ex.ToString());
        // Still return some data
        ret = GetProductMocks();
      }

      return ret;
    }

    protected ObservableCollection<Product> GetProducts(string fileName)
    {
      ObservableCollection<Product> ret;
      XElement elem = XElement.Load(fileName);

      var items = 
        from prod in elem.Descendants("Product")
        select new Product
        {
         ProductId = Convert.ToInt32(prod.Element("ProductId").Value),
         ProductName = prod.Element("ProductName").Value,
         Price = Convert.ToDecimal(prod.Element("Price").Value),
         IsActive = Convert.ToBoolean(prod.Element("IsActive").Value)
        };

      ret = new ObservableCollection<Product>(items);

      return ret;
    }
    #endregion

    #region Insert Method
    public int Insert(Product entity)
    {
      int ret = 0;

      // TODO: Insert Data Here

      return ret;
    }
    #endregion

    #region Update Method
    public int Update(Product entity)
    {
      int ret = 0;

      // TODO: Update Data Here

      return ret;
    }
    #endregion

    #region Delete Method
    public int Delete(Product entity)
    {
      int ret = 0;

      // TODO: Delete Data Here

      return ret;
    }
    #endregion

    #region Mock Data
    protected virtual ObservableCollection<Product> GetProductMocks()
    {
      Product entity;
      ObservableCollection<Product> ret = new ObservableCollection<Product>();

      entity = new Product();
      entity.ProductId = 1;
      entity.ProductName = "(Mock) Product 1";
      entity.Price = 100;
      entity.IsActive = true;
      ret.Add(entity);

      entity = new Product();
      entity.ProductId = 2;
      entity.ProductName = "(Mock) Product 2";
      entity.Price = 200;
      entity.IsActive = true;
      ret.Add(entity);

      entity = new Product();
      entity.ProductId = 3;
      entity.ProductName = "(Mock) Product 3";
      entity.Price = 300;
      entity.IsActive = false;
      ret.Add(entity);

      return ret;
    }
    #endregion
  }
}
